pkg/cursor/pulsar_sub.go

```diff
@@ -0,0 +1,80 @@
+package cursor
+
+import (
+	"context"
+	"time"
+
+	"github.com/apache/pulsar-client-go/pulsar"
+)
+
+var _ Tracker = (*PulsarSubscriptionTracker)(nil)
+
+func NewPulsarSubscriptionTracker(client pulsar.Client, topic string) (*PulsarSubscriptionTracker, error) {
+	consumer, err := client.Subscribe(pulsar.ConsumerOptions{
+		Name:             "pulsar-subscription-tracker",
+		Topic:            topic,
+		SubscriptionName: topic + "-cursor-consumer",
+		Type:             pulsar.Exclusive,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &PulsarSubscriptionTracker{
+		client:   client,
+		consumer: consumer,
+		ch:       consumer.Chan(),
+	}, nil
+}
+
+type PulsarSubscriptionTracker struct {
+	client   pulsar.Client
+	consumer pulsar.Consumer
+	ch       <-chan pulsar.ConsumerMessage
+}
+
+func (p *PulsarSubscriptionTracker) read(ctx context.Context) (cp Checkpoint, err error) {
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	msg, err := p.consumer.Receive(ctx)
+	if err != nil {
+		return cp, err
+	}
+	return ToCheckpoint(msg)
+}
+
+func (p *PulsarSubscriptionTracker) Last() (Checkpoint, error) {
+	var last Checkpoint
+	if p.consumer != nil {
```

Comment:
We can assume that p.consumer is always set.

Reply:
updated in https://github.com/rueian/pgcapture/pull/15/commits/0340fce51c0fd62db0d5665961438b70cd373fc7.